cuda_core/style-guide.md

```diff
@@ -0,0 +1,1980 @@
+
+```python
+cdef bint _initialized = False
+_lock = threading.Lock()
+
+def initialize():
+    global _initialized
+    with _lock:
+        if not _initialized:
+            # Perform initialization
+            _initialized = True
+```
+
+### Guidelines
+
+1. **Use thread-local storage for per-thread state**: When state needs to be isolated per thread, use `threading.local()`.
+
+2. **Protect shared mutable state**: Use locks to protect any shared mutable state that could be accessed from multiple threads.
+
+3. **Minimize lock scope**: Keep critical sections as short as possible to reduce contention.
+
+4. **Document thread safety**: Clearly document which operations are thread-safe and which require external synchronization.
+
+5. **Avoid global mutable state**: Prefer thread-local storage or instance variables over global mutable state when possible.
+
+6. **Combine locks with `nogil` correctly**: Acquire locks before entering `nogil` contexts, not inside them.
+
+## Cython-Specific Features
+
+### Function Declarations
+
+Cython provides three types of function declarations:
+
+1. **`def`**: Python function, callable from Python, slower than C functions
+2. **`cdef`**: C function, not callable from Python, fastest
+3. **`cpdef`**: Hybrid function, callable from both Python and C, faster than `def` but slower than `cdef`
+
+**Guidelines:**
+
+- Use `cdef` for internal helper functions that are only called from Cython code
+- Use `cpdef` when a function needs to be callable from Python but performance is important
+- Use `def` for public Python API functions where flexibility is more important than performance
+
+```python
+# Internal helper - only used in Cython
+cdef inline void Buffer_close(Buffer self, stream):
+    # ...
+
+# Public API - callable from Python, performance important
+cpdef inline int _check_driver_error(cydriver.CUresult error) except?-1 nogil:
+    # ...
+
+# Public API - standard Python function
+def allocate(self, size_t size, stream=None) -> Buffer:
+    # ...
+```
+
+### Class Declarations
+
+Use `cdef class` for Cython extension types:
+
+```python
+cdef class Buffer:
+    cdef:
+        uintptr_t _ptr
+        size_t _size
+        MemoryResource _memory_resource
+```
+
+### The `nogil` Context
+
+Use `nogil` to release the Global Interpreter Lock (GIL) for performance-critical C operations. See [CUDA-Specific Patterns](#cuda-specific-patterns) for detailed guidelines.
+
+### Exception Handling
+
+Use `except?` or `except` clauses to propagate exceptions from `nogil` functions:
+
+```python
+cdef int get_device_from_ctx(...) except?cydriver.CU_DEVICE_INVALID nogil:
+    # Returns CU_DEVICE_INVALID on error, otherwise raises exception
+```
+
+### Type Declarations
+
+Declare C types explicitly for performance:
+
+```python
+cdef:
+    int device_id
+    size_t buffer_size
+    cydriver.CUdeviceptr ptr
+```
+
+### Inline Functions
+
+Use `inline` for small, frequently-called functions:
```

Comment:
Does this still matter in 2025? I am pretty dubious (but of course open to being wrong) that a modern C compiler is going to produce meaningfully faster code versus leaving it to choose whether to inline.